ecosystem/typescript/sdk/src/aptos_client.ts

```diff
@@ -661,6 +661,86 @@ export class AptosClient {
     const txnHash = await this.generateSignSubmitTransaction(sender, payload, extraArgs);
     return this.waitForTransactionWithResult(txnHash, extraArgs);
   }
+
+  /**
+   * Rotate an account's auth key. After rotation, only the new private key can be used to sign txns for
+   * the account.
+   * WARNING: You must create a new instance of AptosAccount after using this function.
+   * @param forAccount Account of which the auth key will be rotated
+   * @param privateKeyBytes New private key
+   * @param extraArgs Extra args for building the transaction payload.
+   * @returns PendingTransaction
+   */
+  async rotateAuthKeyEd25519(
+    forAccount: AptosAccount,
+    privateKeyBytes: Uint8Array,
+    extraArgs?: {
+      maxGasAmount?: BCS.Uint64;
+      gasUnitPrice?: BCS.Uint64;
+      expireTimestamp?: BCS.Uint64;
+    },
+  ): Promise<Gen.PendingTransaction> {
+    const { sequence_number: sequenceNumber, authentication_key: authKey } = await this.getAccount(
+      forAccount.address(),
+    );
+
+    const helperAccount = new AptosAccount(privateKeyBytes);
+
+    const challenge = new TxnBuilderTypes.RotationProofChallenge(
+      TxnBuilderTypes.AccountAddress.CORE_CODE_ADDRESS,
+      "account",
+      "RotationProofChallenge",
+      BigInt(sequenceNumber),
+      TxnBuilderTypes.AccountAddress.fromHex(forAccount.address()),
+      new TxnBuilderTypes.AccountAddress(new HexString(authKey).toUint8Array()),
+      helperAccount.pubKey().toUint8Array(),
+    );
+
+    const challengeHex = HexString.fromUint8Array(BCS.bcsToBytes(challenge));
+
+    const proofSignedByCurrentPrivateKey = forAccount.signHexString(challengeHex);
+
+    const proofSignedByNewPrivateKey = helperAccount.signHexString(challengeHex);
+
+    const payload = new TxnBuilderTypes.TransactionPayloadEntryFunction(
+      TxnBuilderTypes.EntryFunction.natural(
+        "0x1::account",
+        "rotate_authentication_key_ed25519",
+        [],
+        [
+          BCS.bcsSerializeBytes(proofSignedByCurrentPrivateKey.toUint8Array()),
+          BCS.bcsSerializeBytes(proofSignedByNewPrivateKey.toUint8Array()),
+          BCS.bcsSerializeBytes(forAccount.pubKey().toUint8Array()),
+          BCS.bcsSerializeBytes(helperAccount.pubKey().toUint8Array()),
+        ],
+      ),
+    );
+
+    const rawTransaction = await this.generateRawTransaction(forAccount.address(), payload, extraArgs);
+    const bcsTxn = AptosClient.generateBCSTransaction(forAccount, rawTransaction);
+    return this.submitSignedBCSTransaction(bcsTxn);
+  }
+
+  /**
+   * Lookup the original address by the current authKey
+   * @param authKey
+   * @returns original address
+   */
+  async lookupAddressByAuthKey(authKey: MaybeHexString): Promise<HexString> {
```

Comment:
is there a reason we are reintroducing the idea of "authKeys" to clients? When we really want them to put in their address here?

Reply:
This is another conversation probably. But now that we have rotation in the sdk, there is probably a better way to do all these things. For example, encouraging the use of `AptosAccount(private_key)` is probably wrong. We should have the default constructor of AptosAccount's use `lookupAddressByAuthKey`.

Otherwise we are making every developer understand key rotation and understand the necessity of calling `lookupAddressByAuthKey` before calling `AptosAccount(private_key, address)`

Reply:
imo either address or auth key is fine here, address is the same as the auth key on-chain but address is probably less confusing to users? since other chains probably don't have the idea / concept of auth key

Reply:
re: 

> How are wallets going to handle being signed into 2 wallets. Wallet A, rotates keys. Wallet B is now stuck in an unrecoverable state.

I was thinking that it'd be the same as the classic web 2 flow - e.g., when you change your password, you get logged out of everywhere you're signed into this account. And you'll need to sign in with the new password (in this case, the new private key?). lmk if I am missing things tho, not very familiar with the wallet flow here

Reply:
> What error are they going to receive telling them they are logged out? Is it going to be on the first transaction?

I was thinking that it'd be the same as the classic web 2 flow - e.g., when you change your password, you get logged out of everywhere you're signed into this account. And you'll need to sign in with the new password (in this case, the new private key?). lmk if I am missing things tho, not very familiar with the wallet flow here

Reply:
something like `the key of this account has been rotated, pls enter the new key to verify that you're the account owner`? I think on the first transaction is good, or depending on how this is handled on the sdk / wallet we can immediately log them out and ask them for the new key upon a key rotation event (we're adding this soon). what do you think?

Reply:
Sounds good. But I think the error coming from on-chain will just be `Wrong signature` rather than anything helpful